teps/0011-redirecting-step-inputs-and-outputs.md

```diff
@@ -0,0 +1,143 @@
+status: proposed
+---
+
+# TEP-0011: Redirecting Step Inputs and Outputs
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Requirements](#requirements)
+- [Proposal](#proposal)
+    - [Example Usage](#example-usage)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+
+Consuming outputs of a step in another step is a common pattern in writing Tasks. However, this is currently tedious to do. Task authors have to overwrite the image entrypoint with either `sh -c` or the `script` field to wrap up the command to run with an explicit I/O redirection. This is not even possible if the image does not come with a shell, e.g., images built by [`ko`](https://github.com/google/ko).
+
+To achieve the functionality of I/O redirection between steps and even Tasks, Tekton will add new fields to steps for Task authors to specify paths to store stdout/stderr to and read stdin from.
+
+## Motivation
+
+Allow Task authors to support the following use cases:
+
+* Run image `gcr.io/k8s-staging-boskos/boskosctl` with args in a step and process its output through another `jq` image in a subsequent step to acquire the project name without overwriting the image entrypoint, building custom image with both utilities, or looking into container logs ([pipeline#2925](https://github.com/tektoncd/pipeline/issues/2925)).
+
+* Run image `alpine/git` with args `["describe", "--tags"]` in a step and extract the stdout as a Task result without overwriting the image entrypoint to be used by subsequent tasks.
+
+* In general, enable Task authors to access stdout/stderr of a step easily in another step in the same or different Task, without requiring authors to: 1) overwrite image entrypoints; 2) write wrapping scripts solely to do I/O redirection; and 3) build wrapper images to include all needed utilities.
+
+By enabling redirection between steps and Task results, it is much easier for Task authors to composite utility images in a Task.
+
+### Goals
+
+1. Allow the stdout/stderr of a step to be consumed by another step through stdin.
+1. Allow a step to produce a Task result through its stdout or stderr.
+
+### Non-Goals
+
+1. Parse stdout of a step into a structured format (e.g., JSON) and extract information from certain fields.
+
+## Requirements
+
+* Add new fields for a step to specify a file to read from as stdin and files to write to as stdout/stderr.
+* Enhance the `entrypoint` binary to redirect stdin from and stdout/stderr to specified files.
+* Provide examples to use step redirection for the use cases mentioned in [Motivation](#motivation).
+
+## Proposal
+
+The following new fields will be added to the `Step` struct:
+
+```go
+type Step struct {
+    ...
+    // Path to redirect stdin from on container's local filesystem.
+    // +optional
+    Stdin string `json:"stdin,omitempty"`
```

Comment:
Can we add more detail on how `stdin` would work ? (below)

Reply:
I do see some value of having `stdin` but I do think it's not as needed as `stdout` and `stderr`.

Reply:
Also, `stdin`, `stdinOnce` exists in the [`containerSpec`](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/#container-v1-core) and we will also need to take into account `tty`.

Reply:
+1 to dropping this for now. I'm a bit hazy on how stdin will work, especially how it will be more useful than just reading a file, since we won't be able to take advantage of pipes since steps always execute sequentially -- i.e., you won't be able to have step 1 stream data to step 2, since step 1 will (currently) always finish before step 2. So you might as well just write and close a file, then open and read it.

Reply:
We could rename these new fields to avoid conflicts with `containerSpec`.

I'm fine dropping this for now. The only drawback I can think of is that people won't be able no run an image built by `ko` whose command doesn't provide a way to read input from a file.

Reply:
`ko` usage isn't very widespread, especially among containers likely to be used as steps in Tekton, so I wouldn't go too far out of our way to support that use case. If there is some image that operates this way (I think kaniko might be one?) we can work around it either by using an image that has a shell, or by building our own on top that provides a shell.

Reply:
Also, it doesn't prevent image build by `ko` if the base image used (either the default one or the overridden one) has a shell in :upside_down_face:

Reply:
That's true. IMO the inconvenience happens if you are using a third-party image and you cannot/don't want to maintain your own fork.

Also, in my use case, we're building a CI solution based on Tekton, and we don't want to dictate/restrict how end users build their images.

Reply:
Removing stdin redirection from TEP.